### ISSUE TITLE
Generation of a digital receipt

### ISSUE THREAD

**Issue by bjadel** [btcpayserver/btcpayserver#1256]
**Is your feature request related to a problem? Please describe.**
If you run a point of sale it might be legally necessary to offer a digital receipt to the customer.

**Describe the solution you'd like**
Generation of a digital receipt in the form of a (signed) PDF, which can be downloaded via QRCode. This could be implemented as btcTransmuter service. When a new invoice is created, then generate a digital receipt. In the GUI a QRCode could be displayed for a successful transaction, which the user can use to download the receipt if she/he likes. 

In Germany the receipt has to fulfill the following requirements:
https://www.gesetze-im-internet.de/kassensichv/__6.html

The question is what is needed in other countries.

**Comment by bolatovumar** [comment]
Same feature request was discussed on the Lightning dev mailing list. One problem with providing digital receipts is that every country has their own requirements for what receipts must contain.

**Comment by bjadel** [comment]
> Same feature request was discussed on the Lightning dev mailing list. One problem with providing digital receipts is that every country has their own requirements for what receipts must contain.

Yeah, that's right. That's why a very configurable variant would also be important.

**Comment by pavlenex** [comment]
What we can do maybe is show a print view in `Invoice` and perhaps allow field selection (we do something similar for `Payment Request` so for each invoice you'd have a custom field to enter data you need and then select the fields you want included, it's very hard to make this part customizable, so if you have any real receipts it would be helpful to see what kind of information you need.

**Comment by woutersamaey** [comment]
What if we create a text field supporting HTML and placeholders? Similar to an email template system? That way people can customize the receipt as per their countries requirement. Some default or presets could be provided for common countries?

**Comment by pavlenex** [comment]
@woutersamaey That's good idea. Also when invoice is paid, if user uses print-mode of some sort, we can show print icon at the successfully paid invoice at the checkout so they can click right away and print it out if needed.

**Comment by Kukks** [comment]
Use transmuter for this instead of cramming all these features in there.. Currently you can send an email receipt but we can extend it to show a page if you want. it also already handles html + templating

**Comment by nvk** [comment]
Hello coming to comment from twitter thread. Most customers buying from a business account will need to print the invoice, some BTCPayServer are not running emails correctly for receipt.

I propose something as simple as a HTML print call on the page. And if you want to get fancy, a printer friendly css style format with proper invoice headings would be nice.

![image](https://user-images.githubusercontent.com/797193/72534665-30274400-3845-11ea-8cd6-4c4f03e082f6.png)

**Comment by NicolasDorier** [comment]
I think letting merchant upload their own HTML page for the invoice make sense.
I am worried about JS injection thought. A malicious merchant could run javascript on the admin browser.

**Comment by woutersamaey** [comment]
@NicolasDorier just strip all <script>.*</script> and <script.*?> tags from the HTML for security reasons. This should be OK and understood by most.
Why would you need JS on a print view exactly?

**Comment by nvk** [comment]
You can also just link to a new “printer friendly page”, doesn’t need to be JS.

Ps: don’t care about implementation as long as a user can print an invoice :)

**Comment by bolatovumar** [comment]
@NicolasDorier @nvk i guess just printing the completed invoice page with some print-specific CSS as nvk suggests could be a good MVP. If we want to get fancy with that feature down the line then we can always do that later.

**Comment by Transisto** [comment]
I've been asked to merge my issue into this one

> The current invoice format is missing many potentially useful details.
> 
> If a customer doesn't bookmark the invoice and imputing an email are not required there is no way to reliably retrieve an invoiceID if there is any problems.
> 
> The receiving bitcoin address disappear when it's paid, (could be useful to verify confirmations status.
> 
> There is no Bitcoin TXID when a transaction is pending.
> 
> There is no way to monitor confirmation status from the invoice page. (most services do not do any action unless there at least one confirmation. The delay between PAID and ACTUALLY PAID, can be way too long if fee set is too low, it's setting expectation too high but create frustration.

First is there any reason why the address get removed completely from the invoice when something hit the mempool?

There should be a way to display the overly long invoiceID on the invoice.
![image](https://user-images.githubusercontent.com/388716/72691105-c92db780-3af0-11ea-9735-141c181b4cf1.png)
Why do these invoiceID have to be so long? It wouldn't help customer support if these would have to be spelled out loud given how long and cryptic they are.

Most merchant who provide automated services won't trigger action unless the transaction is confirmed, I think it would alleviate a lot the complexity and support of merchant if the invoice could provide a way for customers to monitor whether their tx is confirmed or not.

Bitcoin is a new thing and people what to hold onto and piece of evidence they can in case their transaction doesn't go as planed. Having reusable data like InvoiceID, with address sent to  and Bookmark this, print this page would make a lot feel much more secure.

**Comment by NicolasDorier** [comment]
@Transisto the reason for those Id is pure backward compatibility.

Note that the search field in `Invoices` page support partial matching, so you can just ask your user to give you the first letters.

@bolatovumar actually I think we can just remove the print specific part for now and just having a page like the invoice details page. It renders nicely enough both on screen and on printer.

The data shown would be different though.

**Comment by pavlenex** [comment]
Related slightly to #2621

**Comment by ndeet** [comment]
Some thoughts from dev call just to keep it somewhere:

- show non-sensitive but useful information: amount in fiat, exchange rate, date and time of payment settlement, order id (for POS and crowdfunding there may be different important data) 
- public url, not easily bruteforcable guessable because of random invoice id, but because public sharable similar to shared docs on google docs
- ? directly show the receipt on where we currently display the big green checkmark
- ? show link below return to store button 
- predictable url path like INVOICEID/receipt so integrations can provide the link easily

**Comment by dstrukt** [comment]
Reposting the cleaned up mocks with their variants here for tracking:

<img width="579" alt="Screen Shot 2022-06-09 at 2 12 22 AM" src="https://user-images.githubusercontent.com/6250771/172811433-3c965286-bb27-43ea-a56b-2d8208d0e826.png">
<img width="1426" alt="Screen Shot 2022-06-09 at 2 12 37 AM" src="https://user-images.githubusercontent.com/6250771/172811452-8c4b50fb-e826-478f-84f5-0cb04d5e935f.png">
<img width="1399" alt="Screen Shot 2022-06-09 at 2 12 46 AM" src="https://user-images.githubusercontent.com/6250771/172811461-e5423779-f75e-4a5f-8e75-b2461135608f.png">
<img width="1341" alt="Screen Shot 2022-06-09 at 2 13 09 AM" src="https://user-images.githubusercontent.com/6250771/172811465-7ed4f5d2-73da-4d40-9320-466782f9c46a.png">
<img width="1271" alt="Screen Shot 2022-06-09 at 2 13 20 AM" src="https://user-images.githubusercontent.com/6250771/172811466-21f294b0-990e-429a-b1b2-8c25314b4177.png">
<img width="1256" alt="Screen Shot 2022-06-09 at 2 13 26 AM" src="https://user-images.githubusercontent.com/6250771/172811470-46814837-c664-4882-9250-7ea4df4aee1e.png">

**Comment by Kukks** [comment]
> Some thoughts from dev call just to keep it somewhere:
> 
> * show non-sensitive but useful information: amount in fiat, exchange rate, date and time of payment settlement, order id (for POS and crowdfunding there may be different important data)

I added payment data as opt-in option. There is also an opt-in for additional data to be presented.

> * public url, not easily bruteforcable guessable because of random invoice id, but because public sharable similar to shared docs on google docs

This conflicts with the predictable URL path request below. 

> * ? directly show the receipt on where we currently display the big green checkmark

Would clutter it up too much and the checkout UI does not have space for it

> * ? show link below return to store button

Agreed.

> * predictable url path like INVOICEID/receipt so integrations can provide the link easily
This is how it is.
